---
title: 'ranacapa: An R package to explore environmental DNA data with exploratory statistics and interactive visualizations'
author: Gaurav S. Kandlikar (1,+), Zachary J. Gold (1), Madeline C. Cowen (1), Rachel  S. Meyer (1), Amanda C. Freise (2), Nathan J.B. Kraft (1), Jordan Moberg-Parker (2), Joshua Sprague (3), David Kushner (3), and Emily E. Curd (1)
fontsize: 11pt
output: 
  word_document:
    reference_docx: rendering-templates/word_template.docx
urlcolor: blue
always_allow_html: yes
bibliography: bibliography.bib
csl: rendering-templates/ecology-letters.csl
---
  
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
load("make-figures.Rdata")
options(digits = 3)
```

\setstretch{1.2}

### Affiliations and Contact Information  
(1) Department of Ecology and Evolutionary Biology, University of California – Los Angeles   
(2) Department of Microbiology and Molecular Genetics, University of California – Los Angeles  
(3) National Park Service     

(+) Corresponding Author  

- Email: gkandlikar@ucla.edu
- Phone: (+1) 952-288-7351
- Mailing Address: Dept. of Ecology & Evolutionary Biology, 621 Charles E. Young Drive S., Los Angeles, CA 90095

**Keywords**: environmental DNA; data visualization; citizen science; community science; shiny; metabarcoding; education; community ecology

\clearpage

\setstretch{1.6}

## Abstract 
```{r child = 'abstract.md'}
```

\clearpage

## Introduction

The targeted amplification and sequencing of DNA that living organisms shed into their physical environment, termed "environmental DNA metabarcoding" or "eDNA sequencing",  is revolutionizing microbiology, ecology, and conservation research [@Taberlet_2012; @Deiner_2017].  Sequencing of environmental DNA extracted from field-collected soil, water, or sediment samples can shed light on a range of questions, ranging from profiling the composition of ancient plant and animal communities [@Pedersen_2014], to motoring populations of rare or endangered species [@Balasingham_2017]. As the cost of eDNA sequencing declines and sample collection techniques become more streamlined (e.g. @Thomas_2018), professional research scientists are increasingly using eDNA sequencing as a platform to partner with a diversity of community partners, including natural resource managers, undergraduate students, and citizen scientists in primary research. However, developing robust and impactful community science programs that engage community partners in all steps of the research process remains a challenge.

eDNA sequencing-based projects work well for community science partnerships because non-experts can be quickly trained to collect samples in the field, and because eDNA sequencing is an exciting framework for research pertinent to disciplines such as medicine, agriculture, ecology, and geography [@Deiner_2017]. Community partners in such programs can have heterogeneous backgrounds, ranging from curious members of the public for whom collecting samples in the field is the first scientific research experience (e.g. University of California’s CALeDNA program, http://www.ucedna.com/), to professional natural resource managers who regularly collaborate with research scientists (e.g. Center for Ocean Solutions’ eDNA project, https://oceansolutions.stanford.edu/project-environmental-dna). A key ingredient to ensure sustained engagement of community partners is that they should be able to participate across multiple stages of the research project, not only in sample collection [@Pandya_2012; @ECSA_2015]. This can be a challenge for eDNA sequencing-based community science programs because although it is relatively easy to train community partners to collect eDNA samples, it is far more challenging to train them to interact with and visualize the results from these studies. Indeed, learning the bioinformatic tools necessary for visualizing and analyzing the large, multidimensional datasets generated in these studies can be difficult for professional researchers (Carey & Papin 2018), let alone for the non-technical audience of community science programs. 

To address this challenge, we created an R package “ranacapa”, at the core of which is a Shiny webapp with which users can visualize and perform simple community ecology analyses on results from eDNA sequencing studies. `ranacapa` complements existing visualization platforms (e.g. Phinch, Phyloseq-Shiny, QIIME2 Viewer), because in addition to interactive visualizations, `ranacapa` includes brief explanations and links to additional educational resources to provide users with an overview of basic data analyses used in eDNA studies. `ranacapa` works with community matrices generated via QIIME () or the Anacapa sequence analysis pipeline (https://github.com/limey-bean/Anacapa), which is used extensively by the CALeDNA program.

In the remainder of this manuscript, we describe `ranacapa` and demonstrate its use by two community science partnerships based at the University of California, Los Angeles (UCLA): first, a collaboration between eDNA researchers and resource managers at the National Park Service, and second, a partnership between community ecology researchers and an undergraduate microbiology course at UCLA. As we show in the Use Cases, empowering community partners to interact with the data and perform simple but insightful community ecology analyses can help make these collaborations more enriching and valuable to both parties.     

## Implementation
At the core of `ranacapa` is a Shiny webapp (@Chang_2018), which is avialable at http://gauravsk.shinyapps.io/ranacapa or with `ranacapa::runRanacapa()`. The package also includes two categories of helper functions (Table 1) that transform user-uploaded taxonomy and metadata tables into `R` objects that can be visualized and analuzed using Phyloseq [@McMurdie_2013] and Vegan [@Oksanen_2018]. `ranacapa` is available for installation on Github or CRAN:

```{r eval = F, echo = T}
devtools::install_github("gauravsk/ranacapa")
install.packages("ranacapa")
```

The `ranacapa` Shiny app allows users to interact with eDNA results through statistical summaries and interactive plots, displayed in the following tabs:

- **Sequencing depth**:  Introduces the potential for variation in sequencing depth among samples and explains the basic logic behind rarefying samples in metagenomics studies. Allows users to rarefy the dataset to a sampling depth (Figure 1). The documentation on this tab also acknowledges recent disagreement regarding the value of rarefying in metabarcoding and eDNA sequencing studies [@McMurdie_2014], and allows users to proceed through the rest of the app without rarefying samples.  

#### Figure 1
```{r, fig.cap="Figure 1: Taxon accumulation curve as shown in the first tab of ranacapa. "}
fig1
```

- **Taxonomy heatmap**: Shows the taxonomy matrix as an interactive heatmap made using `heatmaply::heatmaply()`, where the color of each cell represents the number of times a given taxon was sequenced in a sample (Figure 2).   

#### Figure 2
```{r, fig.cap="Figure 2: Taxonomy heatmap as shown in the `ranacapa` Shiny app. Taxonomy is shown at the Order level in this figure; in the app, users can choose the taxonomic level to show in the heatmap."}
fig2
```

- **Taxonomy barplot**: Shows the taxonomy-by-sample matrix as an interactive barplot (Figure 3).  

#### Figure 3
```{r, fig.cap="Figure 3: Taxonomy barplot as shown in the `ranacapa` Shiny app. Taxonomy is shown at the Order level in this figure; in the app, users can choose the taxonomic level to show in the barplot"}
fig3
```

- **Alpha diversity plots**: Introduces the concept of Alpha diversity as the local diversity measured in a single habitat or sample. Users can plot Alpha diversity as observed taxon richness or as Shannon Diversity per sample, or can group samples according to a variable in the metadata file (Figure 4).   

#### Figure 4
```{r, fig.cap="Figure 4: Alpha diveristy boxplots as shown in the `ranacapa` Shiny app. Users can select the X-axis variable using a dropdown menu in the app."}
fig4
```

- **Alpha diversity statistics**: Allows users to choose a variable from the metadata, and generates an Alpha diversity ANOVA table according to the user-selected variable. The tab also shows the output from  a post-hoc Tukey test. 

- **Beta diversity plots**: Introduces the concept of Beta diversity as the turnover in species composition across habitats (or samples). The tab includes an ordination plot generated by `phyloseq::plot_ordination()` based on an ordination made with `phyloseq::ordinate(., method = "PCoA")`. Points on the PCoA plot are colored according to a user-selected metadata variable (Figure 5). The tab also shows dissimilarity among sites according in a dendrogram generated with Ward’s cluster analysis (`stats::hclust(distance_object, method = "ward.d2")`), where `distance_object` is made using `phyloseq::distance()`. For both figures, users can toggle between using Jaccard and Bray-Curtis dissimilarity. 

#### Figure 5
```{r, fig.cap="Figure 5: PCoA ordination of the samples as shown in the `ranacapa` Shiny app. Users can select the grouping variable with a dropdown menu in the app."}
fig5
```

- **Beta diversity statistics**: Shows results from two statistical tests of species turnover across site: first, a multivariate ANOVA implemented with `vegan::adonis())`, which shows users results from a statistical test of taxon turnover across sites, and second, a statistical test of heterogeneity of variances among samples implemented with `vegan::betadisper()`, which shows results from a statistical test that compares the degree of sample-to-sample variation within habitats (or within other user-selected groups).  


## Operation
`ranacapa` depends on Bioconductor v 3.7, which in turn relies on `R` v 3.5.0. The Shiny app has been tested on Chrome and Firefox on Windows, Mac-OSX, and Ubuntu. 

### Input file structure
The `ranacapa` Shiny app requires two input files. The first requirement is a site-by-species matrix, uploaded either as a rich, dense `.biom` table, or as a tab-separated `.txt` file. Qiime2-generated`.qza` files generated by QIIME2 are not immediately suitable for ranacapa, as they do not contain full taxonomy information. If the site-by-species matrix is uploaded as a `.txt` file, the file should match the specifications of the output files from the Anacapa eDNA sequence analysis pipeline (). Specifically, each row in the `.txt` file must represent a taxonomic identification , and each column (save one) must represent the number of times that ASV appears in each sequenced sample. One column, named `sum.taxonomy` must contain the taxonomic identification, with taxonomic rank separated by a semicolon, as follows: 

```
                                                                        sum.taxonomy Arch_point_1 Arch_point_2 Black_seabass_reef_1 Black_seabass_reef_2
  Annelida;Clitellata;Haplotaxida;Hormogastridae;Hormogaster;Hormogaster castillana            0            0                    0                    0
 Annelida;Clitellata;Haplotaxida;Lumbricidae;Allolobophora;Allolobophora chlorotica            0            0                    0                    0
   Annelida;Clitellata;Haplotaxida;Megascolecidae;Amynthas;Amynthas sp. YN201102-05            0            0                    0                    0
        Annelida;Clitellata;Haplotaxida;Megascolecidae;Amynthas;Amynthas thakhantho            0            0                    0                    0
       Annelida;Polychaeta;Capitellida;Maldanidae;Axiothella;Axiothella rubrocincta            0            0                    0                    0
```

The second requirement is a tab-separated `.txt` file that contains sample metadata. The first column should match the sample names in the taxonomy table; the remaining columns of the metadata file contain sample information for each of the samples in the site-by-species matrix. The metadata should contain categorical variables with two or more categories per variable. A valid metadata file for the taxonomy table above would be as follows:

```
                Sample Sample_or_Control         Island  Protection  SST
         Arch_point_1            Sample Santa Barabara unprotected   17
         Arch_point_2            Sample Santa Barabara unprotected   17
 Black_seabass_reef_1            Sample        Anacapa         MPA   16
 Black_seabass_reef_2            Sample        Anacapa         MPA   16
```

The `ranacapa` function `validate_input_files()` verifies that both the taxonomy table and the metadata files match structural requirements, which are documented in the function help files.

## Use Cases
We expect that researchers with bioinformatic expertise will use the pipeline of their choice to assign taxonomy to eDNA datasets and generate clean taxonomy and metadata files that can be visualized in `ranacapa`. Researchers can share these files with their partners, and emphasize the analyses or visualizations most appropriate to their use case. We now show how `ranacapa` can facilitate authentic communication between researchers and community partners in two settings.

### Use Case 1: How `ranacapa` facilitated a collaboration between eDNA researchers and managers at the National Park Service

We expect the interactive taxonomy heatmap to be especially useful for researchers collaborating with natural resource managers whose efforts are focused a targetted list of rare or invasive taxa. For example, UCLA researchers who partner with resource managers at the Channel Islands National Park Service to assess the potential for eDNA as a biodiversity monitoring tool to supplement expensive and time-intensive visual biodiversity surveys in the Southern California Channel Islands [@Lessios_1996, @Usseglio_2015; @Deiner_2017] use ranacapa to share eDNA sequencing results. For this partnership, resource managers collected and filtered thirty-1L water samples for eDNA analysis at permanent monitoring sites inside and adjacent to protected areas, and research scientists at UCLA performed eDNA sequencing of the mitochondrial 12S and CO1 gene, targeting bony fishes, elasmobranches, and invertebrate taxa. The researchers processed sequences and assigned taxonomy using the Anacapa toolkit, and shared results with the resource managers using the `ranacapa` Shiny app. 

The taxonomy heatmap (Figure 2) was the most valuable visualization to this collaboration, because it allowed the resource managers to filter the large observed species list down to a particular set of key taxa that they regularly monitor. The heatmap showed that this pilot study detected 36 of the 70 key metazoans at the species level, and the remaining 34 at the genus, family, or order level. This indicates that eDNA-based studies can likely supplement ongoing management efforts and provide new insights into the spatial and temporal distributions of these species. The value of `ranacapa` in this scenario was to quickly sort through long species lists generated in by eDNA sequencing to highlight the strengths and weaknesses in using eDNA to monitor diversity in the Channel Islands. The data from this study are packaged as the demo dataset for the `ranacapa` Shiny web-app and are available online at XXX.

### Use Case 2: How `ranacapa` helps undergraduate environmental microbiology students pursue sophisticated microbiome analyses

We expect `ranacapa`'s exploratory visualizations and the accompanying explanations to help researchers introducing ecology concepts and analyses in classrooms or other informal education settings. For example, `ranacapa` was used by a research-based environmental microbiology course at UCLA [@Shapiro_2015], in which students used eDNA metabarcoding to study the impact of a local wildfire on the plant and soil microbial community. The goal of this twenty-week course was to provide undergraduate students an authentic experience in basic microbiology and microbial community ecology research. Over the first ten weeks, the instructional team, which included eDNA research scientists, extracted total DNA from student-collected soil samples and sequenced the ITS2 [@Gu_2013] and 16S SSU RNA [@Caporaso_2012] metabarcoding region to characterize the plant, bacterial, and archaeal communities. The instructional team processed sequences and assigned taxonomy using the Anacapa toolkit. 

Students were introduced to the structure of eDNA sequencing results and were taught to explore data and perform simple statistical analyses using `ranacapa`. A key benefit of using `ranacapa` was that despite having no prior bioinformatics experience, students could begin exploring the biodiversity in their samples in a matter of minutes by using the online instance of Shiny app. This allowed the instructors to focus classroom time on biological questions rather than on troubleshooting bioinformatics problems, as had been the case in previous sessions of the course. The course instructors noted that visualizing eDNA data in `ranacapa` helped students understand the relationships between community profile and the various metadata they had collected in the field. By significantly reducing the time and difficulty in visualizing basic biodiversity patterns, `ranacapa` helped students develop and pursue more sophisticated analyses during the remainder of the course, using most sophisticated tools such as STAMP [@Parks_2014] and PICRUSt [@Langille_2013]. The taxonomy tables and metadata files used in this course are available online at XXXX.

## Summary and Future Directions

Metabarcode sequencing of environmental DNA is becoming a key tool in a wide variety of ecological studies, and results from these studies are of interest to a broad audience.  Our `R` package and Shiny web-app `ranacapa` helps users conduct exploratory analyses and visualizations on eDNA datasets, and is a step toward making data and analyses from eDNA sequencing-based studies more accessible and understandable for a wide range of community research partners. 

We propose three avenues for future work in ranacapa. First, we plan to use `ranacapa` as the primary tool to present eDNA results from hundreds of samples sequenced by the CALeDNA community science program.  Second, `ranacapa` will be a key tool in the upcoming undergraduate curriculum module "Pipeline for Undergraduate Microbiome Analysis", which is being built as an open-source, comprehensive suite of analysis and data visualization tools for undergraduate researchers. Finally, in the long-term, we believe there is great promise in connecting `ranacapa` to packages that connect with APIs of online biodiversity databases (e.g. Taxize, rinat). This will help users explore a much wider range of biodiversity questions, for example, by programmatically asking whether their samples include invasive species that are absent from other nearby sites. In sum, tools like `ranacapa` that allow non-technical audiences to easily interact with results from eDNA sequencing studies have great potential to engage community partners with a wide range of backgrounds and interests in primary research.

## Software availability

- A Shiny web-app, including a dataset generated for demonstrations, is available at https://gauravsk.shinyapps.io/ranacapa  
- `ranacapa` is available for installation at https://github.com/gauravsk/ranacapa  
- Link to source code as at time of publication (F1000Research TO GENERATE)  
- Link to archived source code as at time of publication (F1000Research TO GENERATE)  
- Software license (GPL-3)  
- Data availability   
- Figshare: [DOI]  
- License: CC-BY 4.0

## Author contributions
GSK led the development of ranacapa, with help from MCC. ZJG and EEC provided feedback regarding which analyses and visualization options to include. ZJG performed the MPA eDNA study in collaboration with JS and DK. NK, GSK, EC, and RM collaborated with AF and JMP, who used `ranacapa` in their microbiology undergraduate course. GSK wrote the first draft of this manuscript; all authors contributed to revisions.

## Competing interests
No competing interests were disclosed

## Grant information
GSK and ZJG were supported by the US-NSF Graduate Research Fellowship (DEG No. 1650604). EEC is supported by the CALeDNA program, with funds from University of California President’s Research Catalyst Award (CA-16-376437).

## Acknowledgments
We thank Sabrina Shirazi, Rachel Turba, Chris Dao, and Keith Mitchell for providing feedback on developmental versions of this package. We also thank Mahendra Mariadassau and  Pedro Martinez Arbizu for making their `phyloseq-extended` (https://github.com/mahendra-mariadassou/phyloseq-extended)  and `pairwiseAdonis` (https://github.com/pmartinezarbizu/pairwiseAdonis) packages openly available with a GPL-3 License.

\clearpage

## References